src/bin/train.cc

```diff
@@ -432,12 +440,28 @@ int main(int argc, char * argv[]) {
     float lr_decay = 1.0f;
     const float lr_decay_ratio = config.get<float>("Train.lr_decay_ratio");
 
-    // Train/dev/test loop
     const float dropout_ratio = config.get<float>("Train.dropout_ratio");
     const unsigned max_iteration = config.get<unsigned>("Train.max_iteration");
+
+    const string evaluation_type = config.get<string>("Train.evaluation_type");
     const unsigned eval_interval = config.get<unsigned>(
         "Train.evaluation_interval");
     unsigned long num_trained_samples = 0;
+    unsigned long num_trained_words = 0;
+    unsigned long next_eval_words = eval_interval;
+    unsigned long next_eval_samples = eval_interval;
+    if (evaluation_type == "epoch") {
+      next_eval_samples = eval_interval * corpus_size;
+    } else if (evaluation_type == "sample") {
+      next_eval_samples = eval_interval;
+    }
+    auto next_eval_time = 
```

Comment:
Multiple function call system_clock::now() is redundant and may generate different timestamp for each variable. Just call it only once.